File: backend/src/router.py
"""
router.py — Risk-aware routing for SafeRoad AI
alpha=0.0 → fastest route   alpha=1.0 → safest route

Routes over the Almaty graph whose edges were scored by your risk_model.pkl,
so the safety/speed trade-off reflects the real model.
"""

import numpy as np
import networkx as nx
from typing import Any


def _nearest_node(G: nx.MultiDiGraph, lat: float, lng: float) -> int:
    """Find the graph node closest to (lat, lng)."""
    best_node, best_dist = None, float("inf")
    for node, data in G.nodes(data=True):
        d = (data["y"] - lat) ** 2 + (data["x"] - lng) ** 2
        if d < best_dist:
            best_dist, best_node = d, node
    return best_node


def _travel_time(data: dict) -> float:
    length = data.get("length", 100)
    speed = float(data.get("speed_kph", data.get("speedLimit", 50)))
    speed = max(10.0, speed)
    return max(1.0, (length / 1000.0) / speed * 3600.0)


def _combined_cost(data: dict, alpha: float) -> float:
    time_norm = _travel_time(data) / 30.0 * 50.0
    risk = data.get("riskScore", 50.0)
    return (1.0 - alpha) * time_norm + alpha * risk


def find_route(
    G: nx.MultiDiGraph,
    origin_latlon: tuple[float, float],
    dest_latlon: tuple[float, float],
    alpha: float = 0.5,
) -> dict[str, Any]:
    alpha = max(0.0, min(1.0, float(alpha)))

    orig_node = _nearest_node(G, *origin_latlon)
    dest_node = _nearest_node(G, *dest_latlon)

    if orig_node == dest_node:
        raise ValueError("Origin and destination resolve to the same point.")

    # Set combined cost weights
    for u, v, k, data in G.edges(keys=True, data=True):
        G[u][v][k]["_cost"] = _combined_cost(data, alpha)

    try:
        path_nodes = nx.shortest_path(G, orig_node, dest_node, weight="_cost")
    except nx.NetworkXNoPath:
        raise ValueError("No path found between selected points.")

    features, risks, total_time, total_dist = [], [], 0.0, 0.0

    for i in range(len(path_nodes) - 1):
        u, v = path_nodes[i], path_nodes[i + 1]
        edge_data = min(G[u][v].values(), key=lambda d: d.get("_cost", 9999))

        coords = [
            [G.nodes[u]["x"], G.nodes[u]["y"]],
            [G.nodes[v]["x"], G.nodes[v]["y"]],
        ]

        risk  = edge_data.get("riskScore", 50.0)
        band  = edge_data.get("riskBand", "Medium")
        dist  = edge_data.get("length", 100.0)
        ttime = _travel_time(edge_data)

        risks.append(risk)
        total_time += ttime
        total_dist += dist

        # Color each leg by its risk band so the map renders it directly.
        color = "#16a34a" if risk < 50 else ("#ca8a04" if risk < 80 else "#dc2626")

        features.append({
            "type": "Feature",
            "geometry": {"type": "LineString", "coordinates": coords},
            "properties": {
                "riskScore": risk,
                "riskBand": band,
                "length": round(dist, 1),
                "color": color,
            },
        })

    avg_risk = float(np.mean(risks)) if risks else 50.0
    band = "Low" if avg_risk < 50 else ("Medium" if avg_risk < 80 else "High")

    return {
        "geojson":       {"type": "FeatureCollection", "features": features},
        "totalRisk":     round(avg_risk, 1),
        "totalTime":     round(total_time),
        "totalDistance": round(total_dist),
        "riskBand":      band,
        "nodeCount":     len(path_nodes),
        "alpha":         alpha,
        # Mirror fields under the names the frontend normalizer also accepts:
        "riskScore":     round(avg_risk, 1),
        "distance":      round(total_dist),
        "duration":      round(total_time),
    }

File: lib/server/router.ts
/**
 * router.ts — TypeScript port of router.py (risk-aware routing).
 *
 * alpha = 0.0 -> fastest route, alpha = 1.0 -> safest route.
 * Combined edge cost blends normalized travel time against risk score, then we
 * run Dijkstra over the Almaty graph and return colored GeoJSON plus summary
 * stats in the exact shape the frontend's lib/api.ts already normalizes.
 */

import {
  buildGraph,
  type CityGraph,
  type GraphEdge,
} from './almaty-graph'

export interface RouteResponse {
  geojson: {
    type: 'FeatureCollection'
    features: Array<{
      type: 'Feature'
      geometry: { type: 'LineString'; coordinates: number[][] }
      properties: { riskScore: number; riskBand: string; length: number }
    }>
  }
  // Python-style fields
  totalRisk: number
  totalTime: number // seconds
  totalDistance: number // metres
  riskBand: string
  nodeCount: number
  alpha: number
  // Convenience fields the UI consumes directly
  riskScore: number
  distanceKm: number
  durationMin: number
}

function nearestNode(
  g: CityGraph,
  lat: number,
  lng: number,
): number {
  let best = -1
  let bestDist = Infinity
  for (const n of g.nodes) {
    const d = (n.lat - lat) ** 2 + (n.lng - lng) ** 2
    if (d < bestDist) {
      bestDist = d
      best = n.id
    }
  }
  return best
}

function travelTime(edge: GraphEdge): number {
  const speed = Math.max(10, edge.speedKph)
  return Math.max(1, (edge.length / 1000 / speed) * 3600)
}

function combinedCost(edge: GraphEdge, alpha: number): number {
  const timeNorm = (travelTime(edge) / 30) * 50
  return (1 - alpha) * timeNorm + alpha * edge.riskScore
}

/** Min-heap Dijkstra returning predecessor + chosen-edge maps. */
function dijkstra(
  g: CityGraph,
  source: number,
  target: number,
  alpha: number,
): { prev: Map<number, number>; prevEdge: Map<number, GraphEdge> } | null {
  const dist = new Map<number, number>()
  const prev = new Map<number, number>()
  const prevEdge = new Map<number, GraphEdge>()
  dist.set(source, 0)

  // Simple binary heap of [cost, node].
  const heap: Array<[number, number]> = [[0, source]]
  const push = (cost: number, node: number) => {
    heap.push([cost, node])
    let i = heap.length - 1
    while (i > 0) {
      const p = (i - 1) >> 1
      if (heap[p][0] <= heap[i][0]) break
      ;[heap[p], heap[i]] = [heap[i], heap[p]]
      i = p
    }
  }
  const pop = (): [number, number] => {
    const top = heap[0]
    const last = heap.pop()!
    if (heap.length) {
      heap[0] = last
      let i = 0
      for (;;) {
        const l = 2 * i + 1
        const r = 2 * i + 2
        let s = i
        if (l < heap.length && heap[l][0] < heap[s][0]) s = l
        if (r < heap.length && heap[r][0] < heap[s][0]) s = r
        if (s === i) break
        ;[heap[s], heap[i]] = [heap[i], heap[s]]
        i = s
      }
    }
    return top
  }

  const visited = new Set<number>()
  while (heap.length) {
    const [cost, u] = pop()
    if (visited.has(u)) continue
    visited.add(u)
    if (u === target) break
    for (const edge of g.adj.get(u) ?? []) {
      const nd = cost + combinedCost(edge, alpha)
      if (nd < (dist.get(edge.to) ?? Infinity)) {
        dist.set(edge.to, nd)
        prev.set(edge.to, u)
        prevEdge.set(edge.to, edge)
        push(nd, edge.to)
      }
    }
  }

  if (!prev.has(target) && source !== target) return null
  return { prev, prevEdge }
}

export function findRoute(
  origin: [number, number],
  destination: [number, number],
  alphaInput: number,
): RouteResponse {
  const alpha = Math.max(0, Math.min(1, Number(alphaInput)))
  const g = buildGraph()

  const origNode = nearestNode(g, origin[0], origin[1])
  const destNode = nearestNode(g, destination[0], destination[1])
  if (origNode === destNode) {
    throw new Error('Origin and destination resolve to the same point.')
  }

  const result = dijkstra(g, origNode, destNode, alpha)
  if (!result) throw new Error('No path found between selected points.')

  // Reconstruct path from target back to source.
  const pathNodes: number[] = []
  let cur: number | undefined = destNode
  while (cur !== undefined) {
    pathNodes.unshift(cur)
    if (cur === origNode) break
    cur = result.prev.get(cur)
  }

  const features: RouteResponse['geojson']['features'] = []
  const risks: number[] = []
  let totalTime = 0
  let totalDist = 0

  for (let i = 0; i < pathNodes.length - 1; i++) {
    const edge = result.prevEdge.get(pathNodes[i + 1])
    if (!edge) continue
    const a = g.nodes[edge.from]
    const b = g.nodes[edge.to]
    risks.push(edge.riskScore)
    totalTime += travelTime(edge)
    totalDist += edge.length
    features.push({
      type: 'Feature',
      geometry: {
        type: 'LineString',
        // GeoJSON is [lng, lat]
        coordinates: [
          [a.lng, a.lat],
          [b.lng, b.lat],
        ],
      },
      properties: {
        riskScore: edge.riskScore,
        riskBand: edge.riskBand,
        length: round1(edge.length),
      },
    })
  }

  const avgRisk = risks.length
    ? risks.reduce((x, y) => x + y, 0) / risks.length
    : 50
  const band = avgRisk < 50 ? 'Low' : avgRisk < 80 ? 'Medium' : 'High'

  return {
    geojson: { type: 'FeatureCollection', features },
    totalRisk: round1(avgRisk),
    totalTime: Math.round(totalTime),
    totalDistance: Math.round(totalDist),
    riskBand: band,
    nodeCount: pathNodes.length,
    alpha,
    riskScore: round1(avgRisk),
    distanceKm: round2(totalDist / 1000),
    durationMin: round1(totalTime / 60),
  }
}

function round1(v: number): number {
  return Math.round(v * 10) / 10
}
function round2(v: number): number {
  return Math.round(v * 100) / 100
}
